src/librustc_trans/debuginfo/gdb.rs

```diff
@@ -22,19 +23,13 @@ use std::ptr;
 use syntax::attr;
 
 
-/// Inserts a side-effect free instruction sequence that makes sure that the
-/// .debug_gdb_scripts global is referenced, so it isn't removed by the linker.
-pub fn insert_reference_to_gdb_debug_scripts_section_global(ccx: &CrateContext, builder: &Builder) {
+/// Inserts the .debug_gdb_scripts global into the set of symbols
+/// to be placed in `llvm.used`, so it isn't removed by the linker.
```

Comment:
I have confirmed that the linker, called with `--gc-sections`, will, in fact, remove `.debug_gdb_scripts` (whether or not `!llvm.used` is used, of course). Not only that, but it will *also* gladly remove this section even when used exactly [as directed](https://github.com/rust-lang/rust/issues/41626#issuecomment-298163698) by the gcc manual, and building with gcc and binutils ld!

My conclusion is that, I suppose not completely unexpectedly, this feature is poorly thought out and is not widely supported by toolchains, and Rust's original solution is gross but undeniably effective. Ideally, linker scripts (on ELF platforms) would be updated to do something like...
```
SECTIONS {
  .debug_gdb_scripts 0 (NOLOAD) : {
    KEEP(*(.debug_gdb_scripts))
  }
}
```
... but unfortunately, this doesn't work! Without the directive above, the section shows in objdump as:
```
  1 .debug_gdb_scripts 000000aa  ffff0000  ffff0000  00020000  2**0
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
```
which is of course not what we want because it's LOAD. With `(NOLOAD)`, however, it's another story entirely:
```
  1 .debug_gdb_scripts 000000aa  ffff0000  ffff0000  00020000  2**0
                  ALLOC, READONLY
```
So in a direct contradiction with the ld manual, which [claims](https://ftp.gnu.org/old-gnu/Manuals/ld-2.9.1/html_node/ld_21.html) that:

>  The linker will process the section normally, but will mark it so that a program loader will not load it into memory. [...] The contents of the [...] section will appear in the linker output file as usual.

the contents of the section does not, in fact, appear in the output file at all.

I'd like to be proven wrong but I am afraid it's not possible to implement this feature.